aggregator/shard_test.go

```diff
@@ -21,19 +21,107 @@
 package aggregator
 
 import (
+	"math"
 	"testing"
+	"time"
 
 	"github.com/m3db/m3metrics/metric/unaggregated"
 	"github.com/m3db/m3metrics/policy"
+
 	"github.com/stretchr/testify/require"
 )
 
 var (
 	testShard = uint32(0)
 )
 
-func TestAggregatorShard(t *testing.T) {
+func TestAggregatorShardSetWritableRange(t *testing.T) {
+	testNanos := int64(1234)
+	opts := testOptions().
+		SetEntryCheckInterval(0).
+		SetBufferDurationBeforeShardCutover(time.Duration(500)).
+		SetBufferDurationAfterShardCutoff(time.Duration(1000))
+	shard := newAggregatorShard(testShard, opts)
+	inputs := []struct {
+		cutoverNanos          int64
+		cutoffNanos           int64
+		expectedEarliestNanos int64
+		expectedLatestNanos   int64
+	}{
+		{
+			cutoverNanos:          0,
+			cutoffNanos:           int64(math.MaxInt64),
+			expectedEarliestNanos: 0,
+			expectedLatestNanos:   int64(math.MaxInt64),
+		},
+		{
+			cutoverNanos:          testNanos,
+			cutoffNanos:           int64(math.MaxInt64),
+			expectedEarliestNanos: testNanos - 500,
+			expectedLatestNanos:   int64(math.MaxInt64),
+		},
+		{
+			cutoverNanos:          0,
+			cutoffNanos:           testNanos,
+			expectedEarliestNanos: 0,
+			expectedLatestNanos:   testNanos + 1000,
+		},
+	}
+	for _, input := range inputs {
+		shard.SetWriteableRange(input.cutoverNanos, input.cutoffNanos)
+		require.Equal(t, input.expectedEarliestNanos, shard.earliestWritableNanos)
+		require.Equal(t, input.expectedLatestNanos, shard.latestWriteableNanos)
+	}
+}
+
+func TestAggregatorShardIsCutoff(t *testing.T) {
+	now := time.Unix(0, 12345)
+	shard := newAggregatorShard(testShard, testOptions())
+	shard.nowFn = func() time.Time { return now }
+
+	for _, input := range []struct {
+		latestNanos      int64
+		expectedIsCutoff bool
+	}{
+		{latestNanos: 0, expectedIsCutoff: true},
+		{latestNanos: 12345, expectedIsCutoff: false},
+		{latestNanos: math.MaxInt64, expectedIsCutoff: false},
+	} {
+		shard.latestWriteableNanos = input.latestNanos
+		require.Equal(t, input.expectedIsCutoff, shard.IsCutoff())
+	}
+}
+
+func TestAggregatorShardAddMetricWithPoliciesListShardClosed(t *testing.T) {
 	shard := newAggregatorShard(testShard, testOptions().SetEntryCheckInterval(0))
+	shard.closed = true
+	err := shard.AddMetricWithPoliciesList(testValidMetric, testPoliciesList)
+	require.Equal(t, errAggregatorShardClosed, err)
+}
+
+func TestAggregatorShardAddMetricWithPoliciesListShardNotWriteable(t *testing.T) {
+	now := time.Unix(0, 12345)
+	shard := newAggregatorShard(testShard, testOptions())
+	shard.nowFn = func() time.Time { return now }
+
+	for _, input := range []struct {
```

Comment:
nit: same comment as above, maybe move this above the for loop?

Reply:
Same as above.